pkg/service/oidc/service.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pkg/errors"
+	"github.com/pquerna/otp"
+	"github.com/pquerna/otp/totp"
+	"github.com/sirupsen/logrus"
+	credint "github.com/tbd54566975/ssi-service/internal/credential"
+	"github.com/tbd54566975/ssi-service/internal/did"
+	"github.com/tbd54566975/ssi-service/internal/keyaccess"
+	"github.com/tbd54566975/ssi-service/pkg/server/framework"
+	"github.com/tbd54566975/ssi-service/pkg/service/credential"
+	svcframework "github.com/tbd54566975/ssi-service/pkg/service/framework"
+	"github.com/tbd54566975/ssi-service/pkg/service/oidc/model"
+)
+
+const DefaultCNonceExpiration = 5 * time.Second
+
+type Service struct {
+	// key to use for generating nonce values.
+	key             *otp.Key
+	cNonceExpiresIn time.Duration
+
+	resolver    didsdk.Resolver
+	credService *credential.Service
+}
+
+var _ svcframework.Service = (*Service)(nil)
+
+func (s Service) Type() svcframework.Type {
+	return svcframework.OIDC
+}
+
+func (s Service) Status() svcframework.Status {
+	return svcframework.Status{Status: svcframework.StatusReady}
+}
+
+type Option func(*Service)
+
+func WithCNonceExpiresIn(d time.Duration) Option {
+	return func(service *Service) {
+		service.cNonceExpiresIn = d
+	}
+}
+
+func WithOTPKey(k *otp.Key) Option {
+	return func(service *Service) {
+		service.key = k
+	}
+}
+
+func NewOIDCService(didResolver didsdk.Resolver, service *credential.Service, serviceKey [32]byte, opts ...Option) *Service {
+	s := &Service{
+		cNonceExpiresIn: DefaultCNonceExpiration,
+		resolver:        didResolver,
+		credService:     service,
+	}
+	for _, o := range opts {
+		o(s)
+	}
+	if s.key == nil {
+		key, err := totp.Generate(totp.GenerateOpts{
+			Issuer:      "ssi-service",
+			AccountName: "internal",
+			Period:      uint(s.cNonceExpiresIn / time.Second),
+			SecretSize:  32,
+			Digits:      16,
+			Rand:        rand.Reader,
+			Algorithm:   otp.AlgorithmSHA512,
+			Secret:      serviceKey[:],
+		})
+		if err != nil {
+			panic(err)
+		}
+		WithOTPKey(key)(s)
+	}
+
+	return s
+}
+
+func (s Service) CredentialEndpoint(ctx context.Context, credRequest *model.CredentialRequest) (*model.CredentialResponse, error) {
+	if credRequest.Format == "" {
+		return nil, framework.NewRequestErrorMsg("invalid_request", http.StatusBadRequest)
+	}
+	if credRequest.Format != issuance.JWTVCJSON {
+		return nil, framework.NewRequestErrorMsg("unsupported_credential_format", http.StatusBadRequest)
+	}
+	if credRequest.Proof == nil {
+		return nil, errors.New("proof is required")
+	}
+
+	var subject string
+	var err error
+	switch credRequest.Proof.ProofType {
+	case "jwt":
+		subject, err = s.processProof(ctx, credRequest.Proof)
+		if err != nil {
+			return nil, errors.Wrap(err, "processing proof")
+		}
+	default:
+		return nil, errors.New("proof_type not recognized")
+	}
+
+	serviceResp, err := s.credService.GetCredentialsBySubject(ctx, credential.GetCredentialBySubjectRequest{Subject: subject})
```

Comment:
this logic wasn't too straightforward for me to understand...we're getting all credentials for a subject, and then comparing the object model to see if it matches the passed in cred?

I think we can consider different paths for credential equality. if the entire object needs to match, can we simplify to the `id`? What about a triple of `id, schema, issuer` or `issuanceDate`? `id` is probably sufficient...